ops/relation.py

```diff
@@ -0,0 +1,556 @@
+                string specifying acceptable versions for the service
+                type. The version specification can be in any format that is
+                compatible with the
+                `semver <https://pypi.org/project/semver/>`_ Python package.
+
+        Returns:
+            bool: True if the provider and consumer specification are
+               compatible.
+        """
+        # if consumer has no constraints
+        # compatibility is true by default
+        if not needs:
+            return True
+
+        # if consumer has constraints but provider
+        # has no specification compatibility can not
+        # be determined and is hence false by default
+        if not has and needs:
+            return False
+
+        # By now we know both consumer and provider have a
+        # constraint specification so we check if the
+        # constraint type is the same
+        has_type = self._normalized_type(has)
+        needs_type = self._normalized_type(needs)
+        if has_type != needs_type:
+            return False
+
+        # By now we know consumer and provider have the
+        # same constraint type so we check if the constraints
+        # are further qualified by version specifications
+
+        # If consumer is not qualified, provider and
+        # consumer are compatible by default
+        if not self._has_version(needs):
+            return True
+
+        # If consumer is qualified but provider is not there
+        # is no way to determine compatibility so it is False
+        # by default
+        if not self._has_version(has):
+            return False
+
+        # Both consumer and provider are qualified so we
+        # check compatibility of version
+        spec = semver.SimpleSpec(self._normalized_version(needs))
+        got = semver.Version.coerce(self._normalized_version(has))
+
+        return spec.match(got)
+
+    def _has_version(self, constraint):
+        """Does the constraint have a version qualification.
+
+        Args:
+            constraint: a tuple containing a service type (first member)
+                and optionally a service version (second member)
+
+        Returns:
+            bool: True if a service version is present in constraint.
+        """
+        if len(constraint) == 2 and constraint[1] is not None:
+            return True
+        return False
+
+    def _normalized_version(self, spec):
+        """Remove spaces from version strings.
+
+        Args:
+            spec: a tuple containing two members. The second member being
+                a `semver` version specification.
+
+        Returns:
+            str: a version specification that has spaces removed in
+                order to make it compatible with the `semver` package
+                utilities.
+        """
+        version = spec[1]
+        if ' ' in version:
+            return "".join(version.split())
+        else:
+            return version
```

Comment:
Nit, but could make this:

```
return "".join(version.split()) if ' ' in version else version
```

For consistency, it's worth changing `spec` here to `constraint` from the above method since the signatures are the same and they're called with identical arguments in `_is_compatible`

Reply:
Like that. More pythonic. Will change. Thanks.

Reply:
Both suggestions done.